### PR TITLE
[FIX] website / portal language switcher

### PR DESCRIPTION
**Pull request by fpodoo**
The language switcher link on the page /fr/forum/aide-1 should point to /forum/help-1 instead of /forum/aide-1. That fixes error in Google Search Console, and avoid unnecessary redirects to speed up crawlers.

Code cleanup: use a single mechanism for both alternate hreflang, and language switcher. (instead of injecting two different objects in the QWeb evaluation)

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by rdeodoo** [comment]
Spec wise:

> The language switcher link on the page /fr/forum/aide-1 should point to /forum/help-1 instead of /forum/aide-1

It means that on every page we have to read the record name in all available language? Extra query etc, no?
I am not sure about that, especially if it's for google page speed, let alone our testing suite perf.
Do we really want to worsten the perfs on every page load to improve the one if the user switch lang? Which is something very very rare.

**Comment by fpodoo** [comment]
> It means that on every page we have to read the record name in all available language? Extra query etc, no?

No, it was already done for `<link rel="alternate" hreflang="XX"/>` tags. So no extra steps, I just unified both mechanisms.

**Review by Julien00859 — COMMENTED** [review]
It would have been nice to isolate the changes about not supporting the closest lang in a dedicated commit. Here I'm a bit lost with all those website changes that I don't think I should review. So my review is essentially about `http_routing/ir_http.py` and its `_match` override. You tell us if you need our input about res_lang :) Btw it is nice that we drop the support for that closest lang, it made the code complicated :D

File: addons/http_routing/models/ir_http.py
Comment (fpodoo) [reply addons/http_routing/models/ir_http.py:448]: I can put back the `try: finally:` but I'll check if we really need to change the environment of the request? if we can avoid, it's better.  Do you know if there is a test about that? (because this new implementation is green on runbot without it).

> The comment about the lang from the context is wrong, it is the lang of the connected user
My understanding is that it can be a mix; lang of the user, or default lang of the browser; for example in `odoo/http.py`, it sets `context['lang']` to the brower lang. It's the reason I could not remove `get_nearest_lang`, but the goal would be to clean context and remove the `get_nearest_lang`.

Comment (rdeodoo) [reply addons/http_routing/models/ir_http.py:448]: About the try/finally, I don't think it's really needed, it was defensing programming and can probably be removed: https://github.com/odoo/odoo/pull/78857#discussion_r755301678

File: addons/website_links/tests/test_controller.py
Comment (fpodoo) [reply addons/website_links/tests/test_controller.py:43]: I already kept only the minimum; a test that the link tracker works in multi-langs. (I kept `ru`, but it could have been any other lang). I can remove it, but it's just 5 lines, and a valid test that we don't test elsewhere.

File: addons/website/views/website_templates.xml
Comment (rdeodoo) [inline addons/website/views/website_templates.xml:114]: The comment has to be deleted too, if you remove the related `and website and website.is_public_user()` line.
I'd keep those actually, there no need to process this whole block of code (xml/rendering + python calls) for internal users. It doesn't cost much, does it?

As you prefer: keep both or remove both :+1:

File: addons/website/views/website_templates.xml
Comment (rdeodoo) [inline addons/website/views/website_templates.xml:122]: Same indentation as the openin `t-if`?
```suggestion
            </t><t t-else="">
        <link rel="canonical" t-att-href="alternate_languages[0]['href']"/>
            </t>
```

But I actually don't get why we need to do differently if we have 2 langs or 3+ langs

File: addons/website/views/website_templates.xml
Comment (rdeodoo) [inline addons/website/views/website_templates.xml:117]: Why renaming `hreflang` into `url_code` which now sounds like it's the `url_code` properties while it is not?
Hm.. Checking deeper it seems like it is now the `url_code` which is not correct, it shouldn't be, `url_code` is an input that the user can customize to his will, we can't be sure it will respect the expected format.

Before:
```xml
<link rel="alternate" hreflang="fr-be" href="http://127.0.0.1:8069/fr_BE/jobs/detail/developpeur-experimente-4"/```xml
<link rel="alternate" hreflang="fr-fr" href="http://127.0.0.1:8069/francais/jobs/detail/developpeur-experimente-4"/>
```
After:
```xml
<link rel="alternate" hreflang="fr_BE" href="http://127.0.0.1:8069/fr_BE/jobs/detail/developpeur-experimente-4"/>
<link rel="alternate" hreflang="francais" href="http://127.0.0.1:8069/francais/jobs/detail/developpeur-experimente-4"/>
```

Also, on top of that, it seems like another behavior was lost, if you had only one french lang, it would use `fr` without specifying the country code:
Before:
```xml
<link rel="alternate" hreflang="fr" href="http://127.0.0.1:8069/fr_BE/jobs/detail/developpeur-experimente-4"/```xml
```
After:
```xml
<link rel="alternate" hreflang="fr_BE" href="http://127.0.0.1:8069/fr_BE/jobs/detail/developpeur-experimente-4"/>
```

File: addons/website/views/website_templates.xml
Comment (rdeodoo) [inline addons/website/views/website_templates.xml:121]: The canonical URL has nothing to do with the languages, it could very well be like your website has 2 domains (xxx.odoo.com vs mywebsite.be) and we need to show the real end point to avoid multiple indexation or to simply let crawler know which is the real canonical URL.
I am not sure why this is part of the `else` condition of the `if` condition checking language related stuff

(Also, for tracking purpose: note that if you are on a translated version of the website, the canonical is not the main lang one but the translated one on which you are, which is still working fine)

File: addons/website/views/website_templates.xml
Comment (rdeodoo) [inline addons/website/views/website_templates.xml:1952]:
```suggestion
<template id="language_selector" name="Language Selector">
```
Code leftover?

File: odoo/addons/base/models/res_lang.py
Comment (rdeodoo) [inline odoo/addons/base/models/res_lang.py:67]: Seems overkill.
What if tomorrow we introduce another 2 char controller? (if not already, + all the community modules)
I'd just let the user add it and it's not our problem if there is a conflict, whatever is served will be his fault (it's a very very advanced configuration in debug mode, I am sure they wouldn't mind)
It's not worth the code IMHO.

It's a bit like if you create a /shop page while having ecommerce, it won't be accessible because the controller will be served instead, we don't prevent that, it's not worth it

File: odoo/addons/base/models/res_lang.py
Comment (rdeodoo) [inline odoo/addons/base/models/res_lang.py:224]: Could we please name those instead of using the index? It's really hard to figure without going through the stack to see which one is 3, which one is 1 etc.
```suggestion
        for code, url_code, name, active, *_ in self.get_available():
            if active and (url_code == url_code_param):
                return code
```

File: odoo/addons/base/models/res_lang.py
Comment (rdeodoo) [inline odoo/addons/base/models/res_lang.py:274]: `get_available` is hooked by website and it will alter the result if website is in context (mainly for controllers/rpc), is that correct to use it here?

File: addons/website/models/website.py
Comment (rdeodoo) [inline addons/website/models/website.py:1287]: `default_lang_id` is already part of `language_ids`, no need to add it

File: addons/website/tests/test_crawl.py
Comment (rdeodoo) [inline addons/website/tests/test_crawl.py:116]: To restore?

File: odoo/addons/base/models/ir_actions_report.py
Comment (rdeodoo) [inline odoo/addons/base/models/ir_actions_report.py:464]: Do we need to include `wkhtmltopdf` inside the website lang PR?

File: odoo/addons/base/models/res_lang.py
Comment (rdeodoo) [inline odoo/addons/base/models/res_lang.py:297]: Why is this `mya` needed?
We already make sure of that with the XML data: https://github.com/odoo/odoo/blame/07ff047b9dea098135284fedab3f34bf0e31d11b/odoo/addons/base/data/res_lang_data.xml#L7-L9

File: addons/portal/models/ir_qweb.py
Comment (rdeodoo) [inline addons/portal/models/ir_qweb.py:17]: Nitpicking: keep the trailing `,` to ease the diff should someone add another element later, or maybe even go with the `values['is_html_empty']` notation now that it's the only element

File: addons/project/views/project_sharing_templates.xml
Comment (rdeodoo) [inline addons/project/views/project_sharing_templates.xml:15]: Comment was there for this, remove it too :)
If you are sure it's not needed as it was explicitly added to fix something: https://github.com/odoo/odoo/pull/91485

File: odoo/addons/base/models/res_lang.py
Comment (rdeodoo) [inline odoo/addons/base/models/res_lang.py:291]: Cool :+1: 
But I would not to that in `toggle_active` but in `write` IMHO, both will end up calling `write` but we can't be sure all the active switch are going through the `toggle_active`.

File: addons/website/models/website.py
Comment (rdeodoo) [inline addons/website/models/website.py:878]: Why removing this?
I tested on your branch and it seems to be breaking that:

I am on an alias domain `https://24739484-master.runbot173.odoo.com`
The canonical url is `https://24739484-master-all.runbot173.odoo.com` (see the `-all` part)
image

But I still see the alternate tags

File: addons/website/models/website.py
Comment (rdeodoo) [inline addons/website/models/website.py:901]: lost/broken, see other comment more detailed about it

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Here is a first one getting into the code, I didn't really check the templates change yet and didn't really check the perfs yet either.

There is a few behaviors that were removed or wrongly adapted which are not properly tested in our testing suite.
I'd like to write some tests about it before we merge this one: it will be easier to review and develop to just run the testing suite rather than manually checking again and again.

Note that the assumption I had that we kept `/fr/forum/help-1` instead of `/fr/forum/aide-1` for query count/perf reason was wrong because we already have to do the extra queries to get the translated version for the `rel="alternate"` tag (for the public users).
The extra/multiple languages setup is actually not tested in our website perf testing suite.
Before this PR, there would be an extra query for every extra language on the website.
For tracking purpose: it ends up calling `_get_canonical_url_localized()` with each lang one by one which then does `router.build(rule.endpoint, args)` where `args` contains the ORM record with the `lang` in its `context` which is then read (its display name) in the `to_url()` > `slug()` callstack of the `ModelConverter`.

File: addons/website/views/website_templates.xml
Comment (fpodoo) [reply addons/website/views/website_templates.xml:114]: ok, let's keep both.

File: addons/website/views/website_templates.xml
Comment (fpodoo) [reply addons/website/views/website_templates.xml]: The difference is between 1 and 2+; when we have more than two, we have to pick the right one. (the indentation was confusing, just fixed it)

File: addons/website/views/website_templates.xml
Comment (fpodoo) [reply addons/website/views/website_templates.xml]: thanks, fixed.

File: addons/project/views/project_sharing_templates.xml
Comment (fpodoo) [reply addons/project/views/project_sharing_templates.xml:15]: I removed it because it was buggy; `request.context['lang']` is a code, not an url_code (+nothing guarantees that this lang of the user actually exists for the website). Have to test about portal lang redirection though.

File: odoo/addons/base/models/res_lang.py
Comment (fpodoo) [reply odoo/addons/base/models/res_lang.py:291]: I did that initially, but reverted it back; it's easier to not overwrite write to avoid unpredictable side effects. (especially for tests). At least you can write whatever you want in write, it will not force a short version.

File: addons/website/models/website.py
Comment (fpodoo) [reply addons/website/models/website.py:901]: thanks, I fixed this one.

File: addons/website/models/website.py
Comment (fpodoo) [reply addons/website/models/website.py:878]: I removed it for two reasons:
- I need to return the data for the language selector --> no speed gain
- I don't see any reason to remove alternates on non-canonical URLs; so it's easier to always do it.

Comment (rdeodoo) [reply addons/website/models/website.py:878]: According to the introducing commit https://github.com/odoo/odoo/commit/1781041f13b66dfe91b3230794ffb6b2001d86c1
```
Adapt hreflang, because it:
- must only be present on canonical pages
- must always lead to canonical pages
- should not be set if there is no alternate language
```
While you seem to still respect second point, you don't respect first point anymore.
I did not dig into SEO articles to confirm it was the case but I am pretty sure SEB did this for a good reason.
I'd keep it.

File: addons/website/views/website_templates.xml
Comment (rdeodoo) [reply addons/website/views/website_templates.xml:1952]: Never saw that, it will be the first one in the code base, I guess it works but it's really strange and unusual.
Why not just write a migration script as it is supposed to be?
But actually, it should be working migration wise without it, it's not in a `noupdate` section: the XML view should get the update and it should also be replicated on COW views (if not, it's a bug to fix, let me know).

File: odoo/addons/base/models/res_lang.py
Comment (rdeodoo) [reply odoo/addons/base/models/res_lang.py:291]: Fair :+1: , I don't really mind, it's a very advanced feature of a very technical record, let's say it's a bonus when enabling using the toggle_active

File: addons/http_routing/models/ir_http.py
Comment (Julien00859) [reply addons/http_routing/models/ir_http.py]: > but I'll check if we really need to change the environment of the request?

There can be no user yet on the environment (in case of a visitor) but accessing `res.lang` requires querying the database hence we (temporary) grant the public user for those queries. The try/finally is about restoring the original user (maybe `None`, no user) once the queries on `res.lang` are done. In case we remove the try/finally and that there is an exception inside those lines, the environment would still have the public user, it would had leaked. This try/finally is just a common cleanup pattern, it's basically like writing:

```py
with public_user:
    ...
```

But there is no such thing as `with public_user:`, hence the try/finally.

I don't get why you would want to remove that... Ok there is no error atm but then what? A try/finally is basically a no-op when no error occurred. So it's basically free and it is more defensive in case an error actually occurs.

Just... read the comment above "don't try this at home" and trust the fw-py guy please...

> My understanding is that it can be a mix; lang of the user, or default lang of the browser; for example in odoo/http.py, it sets context['lang'] to the brower lang. It's the reason I could not remove get_nearest_lang, but the goal would be to clean context and remove the get_nearest_lang.

You are right, when the user is not logged in (visitor) then we use the lang of the brower to ensure that there always is a lang set in the context. That's an exception from all the other context values that are exclusively taken from the session context, which in turn are set according to the preference set on the user's `res.user` record. So I think it is better to document it that way: lang from the session.

Also I notice it now, `cls._get_default_lang()` uses `tools.get_lang()` which doesn't get the lang from the browser. That comment should be fixed too.

Comment (Julien00859) [reply addons/http_routing/models/ir_http.py]: Edit, I didn't see that the comments were fixed. It is fine :)

**Review by rdeodoo — COMMENTED** [review]
edit: Wrong PR :innocent: (this is being done on another PR)

**Comment by rdeodoo** [comment]
Will be done at https://github.com/odoo/odoo/pull/154242
`url_code` shortener will be done in a separate PR, some things to triple check (`toggle_active` not called anymore when enabling a lang, and possible conflict with `es_419`)

**Pull request closed without merging**
